### DISCUSSION TITLE
VS Code extension combining Scratch + Leopard JS

### DISCUSSION THREAD

**Discussion by cristian-spiescu** [leopard-js/leopard#194] [Show and tell]
Some while ago, I discovered Leopard JS, containing a JS runtime and a JS code generator (from `.sb3` files -> `.js` files). However, proper tooling was not there yet. I was asking about this subject back then (in https://github.com/leopard-js/leopard/discussions/150).

Meanwhile I started working w/ colleagues on a VS Code extension combining Scratch + Leopard JS. I first mentioned it [here](https://github.com/PullJosh/leopard-website/issues/38#issuecomment-2086188196), and then I observed that the "Discussions" section might be a good place to showcase it.

The aim is to be able to combine generated JS (from Scratch) w/ hand written code using OOP/inheritance.

A screenshot:

![image](https://github.com/leopard-js/leopard/assets/4469897/5961c5c1-346c-42df-ab66-506d3dbd2fef)

The project is for the moment very incipient. I'll update when we've got something usable.

**Comment by PullJosh** [comment]
First of all, I love the scratch editor embedded in VSCode. To make sure I'm understanding the project correctly, are you making a scratch project, converting that project to JS, and then writing additional JS code by hand that imports/extends the generated project?

**Reply by cristian-spiescu** [reply]
Exactly. E.g.:

```js
export default class CityBusExt extends CityBus {

    *whenGreenFlagClicked() {
        this.vars.speed = 7;
        this.vars.direction = "e";
        this.x = -xMax;
        this.y = -yMax;
        while (true) {
            fps++;
            switch (this.vars.direction) {
                case "e":
                    this.x += this.toNumber(this.vars.speed);
                    if (this.x > xMax) {
                        this.x = xMax;
                        this.vars.direction = "n";
                    }
                    break;
```
NOTE: if people will use this feature (and/or the things below), this will break compatibility w/ the normal Scratch. 

I will also add some new blocks. 
* First of all a "js block", that can contain arbitrary JS code, e.g. a `console.log()`. This is more an escape hatch.
* Then some more blocks to be able to manipulate variables of types Object. I.e. to allow "communication" between objects (sprites) also via function call.
* Probably also some block to enable functions to return things

**Reply by PullJosh** [reply]
That's fantastic. Lots of people have asked me for two-way syncing between Scratch and JavaScript, and although this isn't that, it seems like you've solving a lot of the same pain-points. Having a JS escape hatch block within Scratch would be cool.

**Comment by sfederici** [comment]
> The project is for the moment very incipient. I'll update when we've got something usable.

Any update?

**Comment by cristian-spiescu** [comment]
Hi 🙂,

Both my colleague and I: we were busy on other projects, so this project was put on hold. However, we restarted work a couple of days ago.

**Reply by sfederici** [reply]
Great news then! I'm eager to have a look at it

**Reply by PullJosh** [reply]
That's awesome! I would love to see more updates as they happen. :)
